### ISSUE TITLE
Exported layer with Values set by Cascading Value Relations tables are converted to strings

### ISSUE THREAD

**Issue by Mikester-Calgary** [qgis/QGIS#45381]
### What is the bug or the crash?

I have cascading Value Relations tables for data entry and they work fine.  Except I need to "Export" the layer in order to keep the displayed values instead of the key values.  Except when I check the boxes that I want to hold the selected values for it changed the data type to string with a length of 254 no matter what the field parameters were set for initially.

### Steps to reproduce the issue

[public_lands_disposition_NAD83_10TM.zip](https://github.com/qgis/QGIS/files/7278771/public_lands_disposition_NAD83_10TM.zip)
[PLA Disp Activities.csv](https://github.com/qgis/QGIS/files/7278777/PLA.Disp.Activities.csv)
[PLA Disp Purpose.csv](https://github.com/qgis/QGIS/files/7278780/PLA.Disp.Purpose.csv)
[PLA Disp Types.csv](https://github.com/qgis/QGIS/files/7278781/PLA.Disp.Types.csv)

[Public Lands Dispositon Data Entry Style.zip](https://github.com/qgis/QGIS/files/7278782/Public.Lands.Dispositon.Data.Entry.Style.zip)
[Test_Output.zip](https://github.com/qgis/QGIS/f
![VR](https://user-images.githubusercontent.com/64865458/135871096-20ed3f30-4036-4fd0-a5e9-308da0d9b46a.png)
iles/7278789/Test_Output.zip)

1: Add public lands shp to project.
2: Style it with the Public Lands style file. qml
3: Notice the field parameters in this layer under Properties ==>Fields.
4: Add a polygon and fill in the drop down fields from the available menu items.
5: Save the shps created.
6: Export the layer to a new shp file retaining the selected values by checking "Replace with displayed values" where appropriate.
7: Notice that all those checked fields in the new file are 254 character strings.

ANY field checked for "Replace with displayed values" gets changed to a 254 character string no matter the incoming field settings.  This can be very damaging when the field parameters are important as they very often are!

### Versions

QGIS version
3.20.3-Odense
QGIS code revision
495fbaec
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.3.1
PROJ version
8.1.1
EPSG Registry database version
v10.028 (2021-07-07)
GEOS version
3.9.1-CAPI-1.14.2
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
13.0
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 2009

Active Python plugins
DataPlotly
db-style-manager
flight_planner
getools
ImportPhotos
kmltools
LAStools
Mergin
mmqgis
processing_wbt
profiletool
qgis2web
QuickPrint
SRTM-Downloader
Three_D_Change_Detection-master
track_profile_2_web
turbo_volume
volume_calculation_tool
db_manager
MetaSearch
processing

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

I have also posted this on StackExchange with no luck so far.

**Comment by gioman** [comment]
> gets changed to a 254 character string

@Mikester-Calgary that is a limit of the shapefile format, use Geopackage instead.

**Comment by Mikester-Calgary** [comment]
Thanks but that is not the issue.  It changes everything to a 254 character string.  10 character strings, Integers, real numbers etc.  It has nothing to do with needing more than 254 characters in the string!  It has to do with completely reformatting the parameters of the field for no reason.  (Well I guess the bug is the reason.)

**Comment by gioman** [comment]
>  [Test_Output.zip](https://github.com/qgis/QGIS/f

@Mikester-Calgary this link is broken

**Comment by Mikester-Calgary** [comment]
[Test_Output.zip](https://github.com/qgis/QGIS/files/7279297/Test_Output.zip)

**Comment by gioman** [comment]
@Mikester-Calgary despite having loaded all the layers when loading the style I still get

```
2021-10-04T16:29:36     WARNING    Missing layer form dependency : layer 'Disposition' requires layer 'PLA Disp Types' to be loaded but 'PLA Disp Types' could not be found, please load it manually if possible.
2021-10-04T16:29:36     WARNING    Missing layer form dependency : layer 'Disposition' requires layer 'PLA Disp Purpose' to be loaded but 'PLA Disp Purpose' could not be found, please load it manually if possible.
2021-10-04T16:29:36     WARNING    Missing layer form dependency : layer 'Disposition' requires layer 'PLA Disp Activities' to be loaded but 'PLA Disp Activities' could not be found, please load it manually if possible.
```

If you can load a QGIS project (together with all the necessary files and styles already loaded) we would appreciate.

**Comment by Mikester-Calgary** [comment]
Those are the csv files that are attached.

**Comment by Mikester-Calgary** [comment]
[AER Entry.zip](https://github.com/qgis/QGIS/files/7279353/AER.Entry.zip)

**Comment by gioman** [comment]
> Those are the csv files that are attached.

@Mikester-Calgary yes, and as I said I still get those message when applying the style, this is why I asked the project.

**Comment by Mikester-Calgary** [comment]
It’s in there.

**Comment by Mikester-Calgary** [comment]
The project is in AER Entry.zip above.

**Comment by gioman** [comment]
> [AER Entry.zip](https://github.com/qgis/QGIS/files/7279353/AER.Entry.zip)

@Mikester-Calgary 

![Screenshot 2021-10-04 163646](https://user-images.githubusercontent.com/1951107/135880934-e245e154-914c-42f2-b0b8-3722ea484714.png)

**Comment by Mikester-Calgary** [comment]
[AER Entry.zip](https://github.com/qgis/QGIS/files/7279385/AER.Entry.zip)

csv files are local in this one.

**Comment by gioman** [comment]
> The project is in AER Entry.zip above.

@Mikester-Calgary the project you sent seems to point to database tables, if you can it would help if you can attach a self contained project, with all the data/layers/styles needed, that opens on any machine without errors.

**Comment by gioman** [comment]
> 6: Export the layer to a new shp file retaining the selected values by checking "Replace with displayed values" where appropriate.
> 7: Notice that all those checked fields in the new file are 254 character strings.

@Mikester-Calgary I was able to fix the project you sent and test your steps, and as I suspected is ok if you save to gpkg, and the fields length changes to 254 if you save to shp.

**Comment by Mikester-Calgary** [comment]
Yes but I can't have the field parameters change!  shp file or otherwise.  Look at the exact parameters of the disposition.shp fields.  The output file needs to have the EXACT same field parameters.  What I am saying is whenever you select "Replace with displayed values" that field gets output as a 254 character string NO MATTER what the field parameters were before export.  If that is intentional that would be ridiculous.
![vr1](https://user-images.githubusercontent.com/64865458/135883523-ca6f477c-280c-4dd9-996a-6dfbf1e52597.JPG)
This image is how the fields are set in the dispositions.shp.  They need to be exported with the exact same parameters as a shp file.

**Comment by Mikester-Calgary** [comment]
![disp_shp](https://user-images.githubusercontent.com/64865458/135883927-0f3e6c9e-01b3-4ea7-9d82-8f1148a4ad3f.JPG)
This is the template.  The previous one is post-export.  Note the field parameter changes.

**Comment by gioman** [comment]
> Yes but I can't have the field parameters change!

@Mikester-Calgary I'm not saying there is no issue here, I'm just adding more details about **when** this happen. And it surely happen if you save as shp, but not as gpkg. This could be an important pointer for a developer that want to give this a look.

**Comment by Mikester-Calgary** [comment]
It seems to me that changing data parameters on export would warrant someone's interest.
